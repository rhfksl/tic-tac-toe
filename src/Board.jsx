import React from "react";

class Board extends React.Component {
  // currentBoard, stepArr 받아서 해당 값 표시
  makeSquare(curBoard, stepArr) {
    return stepArr.map((val) => {
      return (
        <button key={val} className="square" onClick={() => this.props.clickBoard(val)}>
          {curBoard[val]}
        </button>
      );
    });
  }
  render() {
    return (
      <div className="board">
        <div className="board-row">{this.makeSquare(this.props.curBoard, [0, 1, 2])}</div>
        <div className="board-row">{this.makeSquare(this.props.curBoard, [3, 4, 5])}</div>
        <div className="board-row">{this.makeSquare(this.props.curBoard, [6, 7, 8])}</div>
      </div>
    );
  }
}

export default Board;
