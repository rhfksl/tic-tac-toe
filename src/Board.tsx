import React from "react";
import Square from "./Square";
import shortid from "shortid";

interface Props{
  clickBoard(step: number): void;
  curBoard: (string|null)[][];
}

class Board extends React.Component<Props> {
  render() {
    return (
      <div className="board">
        {this.props.curBoard.map((row, rowIdx) => {
          return <Square key={shortid.generate()} curBoard={row} rowIdx={rowIdx} clickBoard={this.props.clickBoard} />;
        })}
      </div>
    );
  }
}

export default Board;
