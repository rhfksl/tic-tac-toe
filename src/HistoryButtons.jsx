import React from "react";

class HistoryButtons extends React.Component {
  render() {
    const historyButtons = this.props.histories.map((_, step) => {
      return (
        <button key={step} onClick={() => this.props.jumpHistory(step)}>
          {step === 0 ? "game start" : `go to ${step} step`}
        </button>
      );
    });
    return (
      <div className="history">
        Jump to history
        <div className="history-button-box">{historyButtons}</div>
      </div>
    );
  }
}

export default HistoryButtons;
