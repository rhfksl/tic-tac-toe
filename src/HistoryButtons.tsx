import React, { useContext } from "react";
import { TicTacToeContext } from './Context/TicTacToeStore';
import { SET_TICTACTOE_STATES } from './Constans/Constans';

export const HistoryButtons: React.FunctionComponent<{}> = () => {
  // const { jumpHistory } = props;
  const { histories, winner, dispatch } = useContext(TicTacToeContext);

  function jumpHistory(newStep: number): void {
    dispatch({ type: SET_TICTACTOE_STATES, payload: { histories: histories, isNextX: newStep % 2 === 0, step: newStep, winner: winner } });
  }

  const historyButtons: JSX.Element[] = histories.map((_, step: number) => {
    return (
      <button key={step} onClick={()=>{jumpHistory(step)}}>
        {step === 0 ? "game start" : `go to ${step} step`}
      </button>
    );
  });

  return (
    <div className="history">
      Jump to history
      <div className="history-button-box">{historyButtons}</div>
    </div>
  );
}