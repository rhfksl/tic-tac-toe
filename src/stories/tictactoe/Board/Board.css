.board{
    width: 50%;
    margin: 100px auto 0px auto;
    display: flex;
    flex-direction: column;
}