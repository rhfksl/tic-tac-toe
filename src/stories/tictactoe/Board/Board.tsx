import React, { useState, useEffect } from "react";
import { RowSquare } from "./Square";
import shortid from "shortid";
import _ from "lodash";

const STRING_O: string = 'O';
const STRING_X: string = 'X';

type Board = (string | null)[][];
interface Traverse{
  (board: Board, row: number, col: number, otherUser: string): null | string;
}

function checkIsWinner (board: Board, row: number, col: number, otherUser: string): null | string{
  // traverse row
    const traverseRow = function (board: Board, row: number, col: number, otherUser: string): null | string {
      for(let i = 0; i < board.length; i++){
        if(!board[row][i] || board[row][i] === '#' || board[row][i] === otherUser){
          return null;
        }
      }
    return otherUser === 'X' ? 'O' : 'X';
  }
  // check col
  const traverseCol = function(board: Board, row: number, col: number, otherUser: string): null | string {
    for(let i = 0 ; i< board.length; i++){
      if(!board[i][col] || board[i][col]=== '#' || board[i][col] === otherUser){
        return null;
      }
    }
    return otherUser === 'X' ? 'O' : 'X';
  }
  
  const traverseClockwiseDiagonal = function (board: Board, row: number, col: number, otherUser: string): null | string {
    
    if(row === col){
      for(let i = 0; i < board.length; i++){
        const boardVal: null | string = board[i][i];
        if(!boardVal || boardVal === otherUser || boardVal === '#'){
          return null;
        }
      }
      return otherUser === 'X' ? 'O' : 'X';
    }
    return null;
  }
  const traverseCounterclockwiseDiagonal = function(board: Board, row: number, col: number, otherUser: string): null | string {
    if(row === board.length - 1 - col){
      for(let i = 0; i < board.length; i++){
        const boardVal: null | string = board[i][board.length - 1 - i];
        if(!boardVal || boardVal === otherUser || boardVal === '#'){
          return null;
        }
      }
      return otherUser === 'X' ? 'O' : 'X';
    }
    return null;
  }

  let helpers: Traverse[] = [traverseRow, traverseCol, traverseClockwiseDiagonal, traverseCounterclockwiseDiagonal];
  
  for (let i = 0; i < helpers.length; i++) {
    let winner: string | null = helpers[i](board, row, col, otherUser);
    if (winner === 'O' || winner === 'X') {
      return winner;
    }
  }
  return null;
}

export interface BoardProps{
    backgroundColor: string;
    boardLength: number;
}

interface BoardStates {
  histories: (null | string)[][][];
  step: number;
  isNextX: boolean;
  winner: (null | string);
}

export const Board: React.FunctionComponent<BoardProps> = props => {
  const { backgroundColor, boardLength } = props;
  const initialBoard: null[][] = [];
  
  for(let i = 0; i < boardLength; i++) initialBoard.push(Array(boardLength).fill(null));

  const [ boardStates, setBoardStates ] = useState<BoardStates>(
    { histories: [ initialBoard ],
      step: 0,
      isNextX: true,
      winner: null,
    }
  );


  useEffect(()=>{
    setBoardStates({...boardStates, histories: [ initialBoard ]})
  }, [boardLength])
  
  const { histories, step, isNextX, winner } = boardStates;
  const curBoard: Board = histories[step];
  
  function clickBoard(boardIdx: number): void{

    if(winner){
      console.log('already winner exists');
      return;
    }
    const cloneHistories: (string|null)[][][] = _.cloneDeep(histories);
    const curBoard: Board = _.cloneDeep(histories[step]);
    const leng: number = curBoard.length;
    const row: number = parseInt(String(boardIdx / leng));
    const col: number = boardIdx % leng;

    if (curBoard[row][col] === STRING_O || curBoard[row][col] === STRING_X) {
      console.log("you already clicked this square");
      return;
    }

    curBoard[row][col] = isNextX ? STRING_X : STRING_O;

    const otherUser: string = isNextX ? STRING_O : STRING_X;
    const isWinner: string | null = checkIsWinner(curBoard, row, col, otherUser);

    setBoardStates({ ...boardStates, histories: [ ...cloneHistories, curBoard ], isNextX: !isNextX, step: step + 1, winner: isWinner } );
  }
    
  return (
          <div className="board">
            <div>current Turn: { isNextX ? STRING_X : STRING_O }</div>
            <div>winner: {winner}</div>
            {curBoard.map((row: (null | string)[], rowIdx: number) => {
              return <RowSquare key={shortid.generate()} row={row} rowIdx={rowIdx} backgroundColor={backgroundColor} clickBoard={clickBoard} />
            })}
          </div>
        );
}

