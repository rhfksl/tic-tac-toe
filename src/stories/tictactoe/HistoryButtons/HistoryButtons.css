button + button{
    margin-left: 0;
}
.history{
    font-size: 15px;
    font-weight: 900;
}

.history-button-box{
    display: flex;
    flex-direction: column;
}

.button-size-small { width: 100px; height: 30px; }
.button-size-medium { width: 150px; height: 50px; }
.button-size-large { width: 200px; height: 80px; }