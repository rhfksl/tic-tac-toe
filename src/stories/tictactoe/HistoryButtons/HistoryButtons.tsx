import React from "react";
import './HistoryButtons.css';

export interface HistoryButtonsProps{
  histories: (null | string)[][][];
  jumpHistory(step: number): void;
  backgroundColor: string;
  color: string;
  size: string;
  fontSize: string;
}

export const HistoryButtons: React.FunctionComponent<HistoryButtonsProps> = props => {
  const { histories, jumpHistory, backgroundColor, color, size, fontSize } = props;

  const historyButtons: JSX.Element[] = histories.map((_, step: number) => {
    return (
      <button 
        key={step} 
        onClick={()=>{jumpHistory(step)}}
        style={{backgroundColor, color, fontSize}}
        className={`button-size-${size}`}
        >
        {step === 0 ? "game start" : `go to ${step} step`}
      </button>
    );
  });

  return (
    <div className="history">
      Jump to history
      <div className="history-button-box">{historyButtons}</div>
    </div>
  );
}