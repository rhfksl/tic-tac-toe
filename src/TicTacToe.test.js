import React from "react";
import renderer from "react-test-renderer";

import isWinner from "./functions";
import TicTacToe from "./TicTacToe";
import Board from "./Board";
import HistoryButtons from "./HistoryButtons";

import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { shallow, mount } from "enzyme";

Enzyme.configure({ adapter: new Adapter() });

describe("Snapshot Check", () => {
  it("check snapshot and TicTacToe component", () => {
    const tree = renderer.create(<TicTacToe />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("<TicTacToe /> Component rendering Check", () => {
  const wrapper = shallow(<TicTacToe />);

  it("renders 1 <Board /> and 1 <HistoryButtons />", () => {
    expect(wrapper.find(Board)).toHaveLength(1);
    expect(wrapper.find(HistoryButtons)).toHaveLength(1);
  });

  it("renders an `#game-main` div element", () => {
    expect(wrapper.find("#game-main")).toHaveLength(1);
  });
});

describe("<TicTacToe /> Component state Check", () => {
  const wrapper = mount(<TicTacToe />);

  it("renders an states correctly", () => {
    const defaultBoard = [
      [
        [null, null, null],
        [null, null, null],
        [null, null, null],
      ],
    ];
    expect(wrapper.state().isNextX).toEqual(true);
    expect(wrapper.state().histories).toEqual(defaultBoard);
    expect(wrapper.state().step).toEqual(0);
  });

  it("renders an 9 squares first", () => {
    expect(wrapper.find(".square")).toHaveLength(9);
  });

  it("clicks first square", () => {
    const square = wrapper.find(".square").first();
    square.simulate("click");
    expect(wrapper.state().isNextX).toEqual(false);
    expect(wrapper.state().step).toEqual(1);
  });

  it("check X is written on first square", () => {
    const square = wrapper.find(".square").first();
    expect(square.text()).toEqual("X");
  });

  it("clicks second square", () => {
    const square = wrapper.find(".square").at(1);
    square.simulate("click");
    expect(wrapper.state().isNextX).toEqual(true);
    expect(wrapper.state().step).toEqual(2);
  });

  it("clicks 0->1->3->4->6 and check winner is X", () => {
    let square = wrapper.find(".square").at(3);
    square.simulate("click");
    square = wrapper.find(".square").at(4);
    square.simulate("click");
    square = wrapper.find(".square").at(6);
    square.simulate("click");
    expect(wrapper.state().isNextX).toEqual(false);
    expect(wrapper.state().step).toEqual(5);
    let histories = wrapper.state().histories;
    expect(isWinner(histories[histories.length - 1])).toBe("X");
  });
});

describe("test isWinner func works well", () => {
  it("when board elements are all null result should be null", () => {
    expect(
      isWinner([
        [null, null, null],
        [null, null, null],
        [null, null, null],
      ])
    ).toBe(null);
  });
  test("when X is winner in 3x3 board", () => {
    expect(
      isWinner([
        [null, "X", "O"],
        ["O", "X", null],
        [null, "X", null],
      ])
    ).toBe("X");
  });
  test("boares are full but nobody wins in 4x4 board", () => {
    expect(
      isWinner([
        ["X", "O", "O", "O"],
        ["O", "X", "X", "X"],
        ["X", "X", "O", "O"],
        ["X", "X", "O", "X"],
      ])
    ).toBe(null);
  });
});
