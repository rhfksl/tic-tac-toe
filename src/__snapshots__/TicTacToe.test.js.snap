// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Check check snapshot and TicTacToe component 1`] = `
<div
  id="game-main"
>
  <div>
    <input
      id="create-board"
      type="text"
    />
    <input
      onClick={[Function]}
      type="submit"
      value="board 생성"
    />
  </div>
  <div
    className="board"
  >
    <div
      className="board-row"
    >
      <button
        className="square"
        onClick={[Function]}
      />
      <button
        className="square"
        onClick={[Function]}
      />
      <button
        className="square"
        onClick={[Function]}
      />
    </div>
    <div
      className="board-row"
    >
      <button
        className="square"
        onClick={[Function]}
      />
      <button
        className="square"
        onClick={[Function]}
      />
      <button
        className="square"
        onClick={[Function]}
      />
    </div>
    <div
      className="board-row"
    >
      <button
        className="square"
        onClick={[Function]}
      />
      <button
        className="square"
        onClick={[Function]}
      />
      <button
        className="square"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="info"
  >
    Tic-Tac-Toe
    <div
      className="history"
    >
      Jump to history
      <div
        className="history-button-box"
      >
        <button
          onClick={[Function]}
        >
          game start
        </button>
      </div>
    </div>
  </div>
  <button
    onClick={[Function]}
  >
    Play
  </button>
  <button
    onClick={[Function]}
  >
    Stop
  </button>
</div>
`;
